src/host/ledger.h

```diff
@@ -871,16 +933,61 @@ namespace asynchost
         require_new_file = true;
       }
 
-      LOG_INFO_FMT("Setting last known/commit index to {}", idx);
       last_idx = idx;
       committed_idx = idx;
+      if (recovery_start_idx_ > 0)
+      {
+        // Do not set recovery idx and create recovery chunks
+        // if the ledger is initialised from 0 (i.e. genesis)
+        recovery_start_idx = recovery_start_idx_;
+      }
+
+      LOG_INFO_FMT(
+        "Set last known/commit index to {}, recovery idx to {}",
+        idx,
+        recovery_start_idx_);
+    }
+
+    void complete_recovery()
+    {
+      // When the recovery is completed (i.e. service is open), temporary
+      // recovery ledger chunks are renamed as they can now be recovered.
+      // Note: this operation cannot be rolled back.
+      LOG_INFO_FMT("Ledger complete recovery");
+
+      for (auto it = files.begin(); it != files.end();)
+      {
+        auto& f = *it;
+        if (f->is_recovery())
+        {
+          f->open();
```

Comment:
Super-corner-case: If there are multiple `.recovery` files in `files` and some renamings fail, we may still end up with a gap in the ledger (`open()` ignores errors). This could lead to gaps, or to missing suffixes, e.g. only the first chunk is renamed, but subsequent ones aren't. Perhaps make `open` check for errors and abort the entire recovery if one of the renamings fails?

Reply:
We should not abort the recovery because one of the nodes is failing its file rename. All our file operations are best effort (host is under control of the attacker) and we hope that at least f+1 nodes are able to perform them correctly.